parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -354,30 +356,44 @@ deepEdits' b = go id b where
     f :: (NameSegment, Branch m) -> Map Name (EditHash, m Patch)
     f (c, b) =  go (addPrefix . Name.joinDot (NameSegment.toName c)) (head b)
 
+data MergeMode = RegularMerge | SquashMerge deriving (Eq,Ord,Show)
+
 merge :: forall m . Monad m => Branch m -> Branch m -> m (Branch m)
-merge b1 b2 | isEmpty b1 = pure b2
-merge b1 b2 | isEmpty b2 = pure b1
-merge (Branch x) (Branch y) =
-  Branch <$> Causal.threeWayMerge combine x y
+merge = merge' RegularMerge
+
+-- Discards the history of a Branch0's children, recursively
+discardHistory0 :: Applicative m => Branch0 m -> Branch0 m
+discardHistory0 b = over children (fmap tweak) $ b where
+  tweak b = cons (discardHistory0 (head b)) empty
+
+merge' :: forall m . Monad m => MergeMode -> Branch m -> Branch m -> m (Branch m)
```

Comment:
This is the main change - there's a `MergeMode` which is either `RegularMerge` or `SquashMerge`, and this is plumbed through.